### PR TITLE
Apply dynamic type in conversation view.

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
This PR is now ready for review.

PTAL @michaelkirk 

This only covers the most common text in conversation view: message bodies, status messages, etc.  I haven't done less common places (like missing attachment indicator) yet.

### CODE REVIEW

**Comment by charlesmchen-signal** [comment]
simulator screen shot - iphone x - 2018-04-06 at 16 40 46
simulator screen shot - iphone x - 2018-04-06 at 16 40 38
simulator screen shot - iphone x - 2018-04-06 at 16 40 33
simulator screen shot - iphone x - 2018-04-06 at 16 39 22
simulator screen shot - iphone x - 2018-04-06 at 16 39 18
simulator screen shot - iphone x - 2018-04-06 at 16 34 38
simulator screen shot - iphone x - 2018-04-06 at 16 34 31

File: Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m [hunk 1]
```diff
@@ -25,18 +25,12 @@ @interface OWSSystemMessageCell ()
 @property (nonatomic) UIImageView *imageView;
 @property (nonatomic) UILabel *titleLabel;
 
-// override from JSQMessagesCollectionViewCell
-@property (nonatomic) UILabel *cellTopLabel;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m]: Obsolete view.

File: SignalMessaging/categories/UIFont+OWS.h [hunk 2]
```diff
@@ -28,11 +26,20 @@ NS_ASSUME_NONNULL_BEGIN
...
 @property (class, readonly, nonatomic) UIFont *ows_dynamicTypeTitle2Font;
+@property (class, readonly, nonatomic) UIFont *ows_dynamicTypeTitle3Font;
 @property (class, readonly, nonatomic) UIFont *ows_dynamicTypeHeadlineFont;
-@property (class, readonly, nonatomic) UIFont *ows_infoMessageFont;
-@property (class, readonly, nonatomic) UIFont *ows_footnoteFont;
+@property (class, readonly, nonatomic) UIFont *ows_dynamicTypeSubheadlineFont;
+@property (class, readonly, nonatomic) UIFont *ows_dynamicTypeBodyFont;
+@property (class, readonly, nonatomic) UIFont *ows_dynamicTypeFootnoteFont;
+@property (class, readonly, nonatomic) UIFont *ows_dynamicTypeCaption1Font;
+@property (class, readonly, nonatomic) UIFont *ows_dynamicTypeCaption2Font;
```
Comment (charlesmchen-signal) [inline SignalMessaging/categories/UIFont+OWS.h]: I'm not sure we'll need all of these, but we now have the canonical list, in correct order.

File: SignalMessaging/categories/UIFont+OWS.h [hunk 3]
```diff
@@ -28,11 +26,20 @@ NS_ASSUME_NONNULL_BEGIN
...
+
+#pragma mark - Styles
+
+- (UIFont *)ows_italic;
+- (UIFont *)ows_medium;
```
Comment (charlesmchen-signal) [inline SignalMessaging/categories/UIFont+OWS.h]: I believe we only ever use medium, not bold.

File: SignalMessaging/categories/UIFont+OWS.m [hunk 4]
```diff
@@ -57,29 +52,85 @@ + (UIFont *)ows_elegantIconsFont:(CGFloat)size
...
+}
+
+- (UIFont *)ows_medium
+{
+    // The recommended approach of deriving "medium" weight fonts for dynamic
+    // type fonts is:
+    //
+    // [UIFontDescriptor fontDescriptorByAddingAttributes:...]
+    //
+    // But this doesn't seem to work in practice on iOS 11 using UIFontWeightMedium.
```
Comment (charlesmchen-signal) [inline SignalMessaging/categories/UIFont+OWS.m]: See this comment.

**Comment by charlesmchen-signal** [comment]
This PR is now ready for review.

PTAL @michaelkirk

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 5]
```diff
@@ -100,12 +100,12 @@ - (void)loadForDisplay
...
 - (UIFont *)titleFont
 {
-    return [UIFont ows_mediumFontWithSize:16.f];
+    return UIFont.ows_dynamicTypeBodyFont.ows_medium;
 }
 
 - (UIFont *)buttonFont
 {
-    return [UIFont ows_regularFontWithSize:14.f];
+    return UIFont.ows_dynamicTypeBodyFont;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: Is this change in size intentional? We're going from 14 -> 17

File: Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m (hunk 1 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m]: This happened a long time ago but I think we agreed to this design change.  I think they should be able to use message details if they really wanted to see when this happened.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 6]
```diff
@@ -1101,7 +1101,7 @@ - (void)setNavigationTitle
         name = [self.contactsManager
             attributedStringForConversationTitleWithPhoneIdentifier:self.thread.contactIdentifier
                                                         primaryFont:[self navigationBarTitleLabelFont]
-                                                      secondaryFont:[UIFont ows_footnoteFont]];
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Good point; made this non-dynamic.

**Pull request merged**